tutor/commands/upgrade/compose.py

```diff
@@ -158,6 +158,34 @@ def upgrade_from_olive(context: click.Context, config: Config) -> None:
     upgrade_mongodb(context, config, "4.2.17", "4.2")
     upgrade_mongodb(context, config, "4.4.22", "4.4")
 
+    upgrade_to_redwood_onwards, new_mysql_docker_image = (
+        common_upgrade.verify_tutor_version_for_mysql_upgrade(config)
+    )
+
+    if not upgrade_to_redwood_onwards:
+        return
+
+    # Revert the MySQL image first to build data dictionary on v8.1
+    old_mysql_docker_image = "docker.io/mysql:8.1.0"
+    click.echo(fmt.title(f"Upgrading MySQL to v{new_mysql_docker_image.split(':')[1]}"))
+    config["DOCKER_IMAGE_MYSQL"] = old_mysql_docker_image
+    # Note that the DOCKER_IMAGE_MYSQL value is never saved, because we only save the
+    # environment, not the configuration.
+    tutor_env.save(context.obj.root, config)
+    context.invoke(compose.start, detach=True, services=["mysql"])
+    fmt.echo_info("Waiting for MySQL to boot...")
+    sleep(30)
```

Comment:
Is there any way to avoid that sleep statement? It's bound to be either too long or too short for many people. We could avoid that if we could run some sort of `mysql --upgrade` one-off command, instead of starting the full mysql service.

Reply:
From these [docs](https://dev.mysql.com/doc/refman/8.0/en/mysql-upgrade.html):

```
As of MySQL 8.0.16, the MySQL server performs the upgrade tasks previously handled by mysql_upgrade.
Consequently, mysql_upgrade is unneeded and is deprecated as of that version; expect it to be removed in a future
version of MySQL. Because mysql_upgrade no longer performs upgrade tasks, it exits with status 0 unconditionally.
```

And this [mysql-8-0-16-mysql_upgrade-is-going-away](https://dev.mysql.com/blog-archive/mysql-8-0-16-mysql_upgrade-is-going-away/) as well, it is not possible to run a one off `mysql upgrade` command. MySQL handles the upgrade process itself whenever a new MySQL server is started and it detects that an upgrade process is required.

The official [guide](https://dev.mysql.com/doc/refman/8.0/en/docker-mysql-getting-started.html#docker-upgrading) on how to upgrade docker based mysql servers also states the same as above.

However, what we can do is healthcheck polling to wait until the server is ready for connections before returning similar to what is being done in the [mysql init job](https://github.com/overhangio/tutor/blob/master/tutor/templates/jobs/init/mysql.sh#L4-L15).

Reply:
poll check with a timeout might be better suited here. In case the service does not start after 30 seconds, there would be no point in executing the remaining operations.